### ISSUE TITLE
General specification of piecewise functions with jumps

### ISSUE THREAD

**Issue by ChristianZimpelmann** [iza-institute-of-labor-economics/gettsim#353]
### Current and desired situation

We currently use parameters of `type: piecewise_linear` to specify parameters of piecewise functions. These parameters are loaded using `get_piecewise_parameters`. We thereby assume that the pieces of the functions are connected at the thresholds and that there are no jumps.

A typical potential use case of a piecewise function is that a parameter can take on a set of values, depending on another variable. E.g. 

```yaml
freib_behinderung:
  name:
    de: Jährlicher Freibetrag pro behindertes Haushaltsmitglied abhängig vom Behinderungsgrad.
  1991-01-01:
    0: 0
    1: 1224
    80: 1536
```

This case cannot be handled by `get_piecewise_parameters` because the function has jumps at 1 and at 80.

### Proposed implementation

- make `get_piecewise_parameters` more flexible
- potentially add a new `type` for functions that are constant in each bin (in this case, no key `rates` is needed)

### Considered alternatives

- create a new function for values that differ based on certain bins of another variable, but stay constant within a bin.

**Comment by hmgaudecker** [comment]
Isn't the last thing something like `numpy.searchsorted`? Wildly guessing here, not sure I understand the use case fully.

**Comment by ChristianZimpelmann** [comment]
Seems like `numpy.searchsorted` would be a good solution after we moved to scalars.

But nevertheless, it would make sense to standardize the parameter definition (and loading of these parameters) in a similar fashion as we do for `type: piecewise_linear` functions. In the example above it is for example not directly clear on which end the intervals are open and if the interval between 1 and 80 is assigned to 1224 or 1536.

**Comment by hmgaudecker** [comment]
Just a reminder that we need to think of ways to accommodate cases where the distinction between >= and > seriously matters, e.g. birth years in `gettsim/parameters/arbeitsl_geld_2.yaml` / `vermögensgrundfreibetrag_obergrenze` should ideally be `bis_1947`, `ab_1948_bis_1957`, etc.

**Comment by hmgaudecker** [comment]
Cross-linking #203

**Comment by hmgaudecker** [comment]
Would a simple reformulation of the problem not help? Or is it a mix of absolute and relative numbers?

**Comment by hmgaudecker** [comment]
Is this still relevant? In any case, I think the title needs adjustment.

- I do not think there will be a version of `piecewise_polynomial` with jumps. I currently relies on being continuous, adding another option would mean having to carry an additional parameter everywhere, this is very unlikely to be worth it.
- I do not quite see which use cases are relevant still for what is left then.

**Comment by ChristianZimpelmann** [comment]
I think it is still relevant. Use cases are e.g. `freib_behinderung` or [Anspruchsdauer for ALG depending on age](https://github.com/iza-institute-of-labor-economics/gettsim/blob/ae1ed32da13cf8179639df82676962e39b2821b7/src/_gettsim/transfers/arbeitsl_geld.py#L74).

- I am open for other ideas how to handle the situations in which the relation between two columns depends on cutoffs, but is not continuous. Currently, we use in these cases `piecewise_polynomial` with an "array of zeros"-hack
- Adding a new function which can handle these situations would also work out. The question is whether that function wouldn't make the current ` piecewise_polynomial` function obsolete (at least if it could handle functions that are non-continuous and are not constant within intervals).

**Comment by hmgaudecker** [comment]
Are we thinking of those examples even as continuous? My intuition would be they are all piecewise constant, which is `np.searchsorted` (I believe; and I am not sure whether there is a Jax alternative).

In any case, I would think a second function, which does not require polynomial degrees beyond 0 (or maybe 1) would be more helpful than trying to handle all cases within one monster.

**Comment by ChristianZimpelmann** [comment]
I agree that all those cases are piecewise constant. So a second function should be fine for now.

**Comment by hmgaudecker** [comment]
I still believe the function is already there in [np.searchsorted](https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html), also [part of jax.numpy](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.searchsorted.html). Should become pretty natural to use once we have a clear protocol for the pre-processing of parameters.

**Comment by hmgaudecker** [comment]
This is a great example of how the pre-processing will help.

- We specify the couple of parameters in a completely natural way in the yaml file
- We have a function taking the evaluation year and the original parameters as inputs, converting those to the value in a given year.

AFAIU, there is nothing here that makes the above dependent on the data, so I am not sure why we need piecewise polynomial in the first place (other than conversion of parameters being awkward/impossible in current structure)?

**Comment by ChristianZimpelmann** [comment]
> We have a function taking the evaluation year and the original parameters as inputs, converting those to the value in a given year.

That is basically the current function `_parse_einführungsfaktor_vorsorgeaufw_alter_ab_2005` (which does not have to be explicitely called of course), isn't it?

No need to use piecewise polynomial, but I am not sure how it could be specified nicer in a "natural way". Could you sketch an example?

**Comment by hmgaudecker** [comment]
Like this?

```py
@dates_active(start_date=2005, end_date=2022)
def einführungsfaktor_vorsorgeaufw_alter(
    aktuelles_jahr, 
    einführungsfaktor_ausgangswert,
    einführungsfaktor_start_jahr,
    einführungsfaktor_rate
):
    faktor = einführungsfaktor_ausgangswert + einführungsfaktor_rate * (
        aktuelles_jahr
        - einführungsfaktor_start_jahr
     )
```

Current implementation seems to have a bit of history dependence in it... Unless I am missing complexity somewhere.

**Comment by ChristianZimpelmann** [comment]
I see, and from 2023 it would be a different parameter in the `yaml`. Makes sense, I will leave the parameter as is for now, but this seems indeed much nicer.

I will try out `np.searchsorted` on next occasion.

**Comment by hmgaudecker** [comment]
I'd even say it would be completely gone from 2023 onwards. Just a different regime, the above is only relevant for the transition period, as the name suggests.

**Comment by hmgaudecker** [comment]
Done with all the changes in #804
